### PR TITLE
assumeReadType for EquiJoin

### LINKED ISSUE #2878: MongoDbPlanner uses mapreduce for joins (phase 1)
MongoDbPlanner does not use `$lookup` when it should.

Query:
```
select extraSmallZips.city from smallZips join extraSmallZips on smallZips.`_id` = extraSmallZips.`_id`
```

Mongo v3.2 planner:

```
db.smallZips.aggregate(
  [
    {
      "$project": {
        "0": {
          "$arrayElemAt": [["$_id", "$$ROOT"], { "$literal": NumberInt("1") }]
        },
        "1": { "$literal": true },
        "src": ["$_id", "$$ROOT"]
      }
    },
    {
      "$match": { "$and": [{ "0": { "$type": NumberInt("3") } }, { "1": true }] }
    },
    { "$out": "tmp.gen_4c74c4e_3" }],
  { "allowDiskUse": true });
db.tmp.gen_4c74c4e_3.mapReduce(
  function () {
    emit.apply(
      null,
      (function (key, value) { return [key, { "0": value.src[1]._id, "content": value.src }] })(
        this._id,
        this))
  },
  function (key, values) { return values[0] },
  { "out": { "replace": "tmp.gen_4c74c4e_2", "db": "test" } });
db.tmp.gen_4c74c4e_2.aggregate(
  [
    {
      "$group": { "0": { "$push": "$value.content" }, "_id": { "0": "$value.0" } }
    },
    {
      "$project": {
        "_id": true,
        "value": { "left": "$0", "right": { "$literal": [] }, "_id": "$_id" }
      }
    },
    { "$out": "tmp.gen_4c74c4e_1" }],
  { "allowDiskUse": true });
db.extraSmallZips.aggregate(
  [
    {
      "$project": {
        "0": {
          "$arrayElemAt": [["$_id", "$$ROOT"], { "$literal": NumberInt("1") }]
        },
        "1": { "$literal": true },
        "src": ["$_id", "$$ROOT"]
      }
    },
    {
      "$match": { "$and": [{ "0": { "$type": NumberInt("3") } }, { "1": true }] }
    },
    { "$out": "tmp.gen_4c74c4e_5" }],
  { "allowDiskUse": true });
db.tmp.gen_4c74c4e_5.mapReduce(
  function () {
    emit.apply(
      null,
      (function (key, value) {
        return [{ "0": value.src[1]._id }, { "left": [], "right": [value.src] }]
      })(
        this._id,
        this))
  },
  function (key, values) {
    var result = { "left": [], "right": [] };
    values.forEach(
      function (value) {
        result.left = result.left.concat(value.left);
        result.right = result.right.concat(value.right)
      });
    return result
  },
  { "out": { "reduce": "tmp.gen_4c74c4e_1", "db": "test", "nonAtomic": true } });
db.tmp.gen_4c74c4e_1.aggregate(
  [
    {
      "$match": {
        "value.left": { "$not": { "$size": NumberInt("0") } },
        "value.right": { "$not": { "$size": NumberInt("0") } }
      }
    },
    { "$unwind": "$value.right" },
    { "$unwind": "$value.left" },
    { "$project": { "value": true, "_id": false } },
    { "$out": "tmp.gen_4c74c4e_0" }],
  { "allowDiskUse": true });
db.tmp.gen_4c74c4e_1.drop();
db.tmp.gen_4c74c4e_5.drop();
db.tmp.gen_4c74c4e_3.drop();
db.tmp.gen_4c74c4e_2.drop();
db.tmp.gen_4c74c4e_0.mapReduce(
  function () {
    emit.apply(
      null,
      (function (key, value) {
        return (function (key, value) { return [key, value.right[1].city] }).apply(
          null,
          (function (key, value) { return [key, value.value] })(key, value))
      })(
        this._id,
        this))
  },
  function (key, values) { return values[0] },
  { "out": { "inline": NumberLong("1") } });
```

Legacy mongo v3.2 planner:

```
db.smallZips.aggregate(
  [
    { "$match": { "_id": { "$exists": true } } },
    { "$project": { "left": "$$ROOT" } },
    {
      "$lookup": {
        "from": "extraSmallZips",
        "localField": "left._id",
        "foreignField": "_id",
        "as": "right"
      }
    },
    { "$unwind": "$right" },
    { "$limit": NumberLong("11") },
    { "$project": { "value": "$right.city", "_id": false } }],
  { "allowDiskUse": true });
```

### PR DESCRIPTION
**Pull request by rintcius**
Fixes #2878

### CODE REVIEW

File: mongodb/src/test/scala/quasar/physical/mongodb/PlannerSql2ExactSpec.scala [hunk 1]
```diff
@@ -70,22 +71,20 @@ class PlannerSql2ExactSpec extends
...
-            free.ShiftedRead[AFile](rootDir </> dir("db") </> file("zips"), qscript.ExcludeId),
-            func.Guard(func.Hole, Type.AnyObject, func.Constant(json.bool(true)), func.Constant(json.bool(false)))),
-          free.Filter(
-            free.ShiftedRead[AFile](rootDir </> dir("db") </> file("smallZips"), qscript.ExcludeId),
-            func.Guard(func.Hole, Type.AnyObject, func.Constant(json.bool(true)), func.Constant(json.bool(false)))),
-          List((func.ProjectKeyS(func.Hole, "_id"), func.ProjectKeyS(func.Hole, "_id"))),
+          free.ShiftedRead[AFile](rootDir </> dir("db") </> file("zips"), qscript.ExcludeId),
+          free.ShiftedRead[AFile](rootDir </> dir("db") </> file("smallZips"), qscript.ExcludeId),
+          List((
+            func.Guard(func.Hole, Type.AnyObject, func.ProjectKeyS(func.Hole, "_id"), func.Undefined),
```
Comment (jsantos17) [inline mongodb/src/test/scala/quasar/physical/mongodb/PlannerSql2ExactSpec.scala]: I think these `Guard`s can be elided as well. It's my understanding `Hole` here refers either to `lBranch` or `rBranch`, which are `ShiftedRead`s with a known type.

Comment (rintcius) [reply mongodb/src/test/scala/quasar/physical/mongodb/PlannerSql2ExactSpec.scala]: Yes, but in this spec the qscript is whatever the outcome is of `sql2 -> qscript` and this is what we get given `sql2 = sqlE"select smallZips.city from zips join smallZips on zips._id = smallZips._id"`. If all guards were properly elided before it reaches mongo then we wouldn't have to elide anything in `assumeReadType` anymore.

Comment (jsantos17) [reply mongodb/src/test/scala/quasar/physical/mongodb/PlannerSql2ExactSpec.scala]: Oh my bad. I thought this was already Mongo-specific QScript and we were not eliding the `Guard`s here. What do you think about applying Mongo-specific transformations to the QScript here? That's what we're going to use for plans anyway.

Probably not in this PR tho.

Comment (rintcius) [reply mongodb/src/test/scala/quasar/physical/mongodb/PlannerSql2ExactSpec.scala]: I took the qscript that's being used in the call to `MongoDbPlanner.plan`. I think that makes a good point to separate what's mongo's responsibility and what's not. But it's not really stable qscript. I'd like to discuss this in upcoming meeting.

**Pull request merged**